blockchain/blockdao/filedao_v2_util.go

```diff
@@ -0,0 +1,246 @@
+}
+
+func (fd *fileDAOv2) putBlock(blk *block.Block) error {
+	blkInfo := &block.Store{
+		Block:    blk,
+		Receipts: blk.Receipts,
+	}
+	ser, err := blkInfo.Serialize()
+	if err != nil {
+		return err
+	}
+	blkBytes, err := compressDatabytes(ser, fd.compressBlock)
+	if err != nil {
+		return err
+	}
+
+	// add to staging buffer
+	index := stagingKey(blk.Height(), fd.bottomHeight)
+	full, err := fd.blkBuffer.Put(index, blkInfo)
+	if err != nil {
+		return err
+	}
+	if !full {
+		fd.batch.Put(stagingDataNS, byteutil.Uint64ToBytesBigEndian(index), blkBytes, "failed to put block")
+		return nil
+	}
+
+	// pack blockStoreBatchSize blocks together, write to block store
+	if ser, err = fd.blkBuffer.Serialize(); err != nil {
+		return err
+	}
+	if blkBytes, err = compressDatabytes(ser, fd.compressBlock); err != nil {
+		return err
+	}
+	return addOneEntryToBatch(fd.blkStore, blkBytes, fd.batch)
+}
+
+func (fd *fileDAOv2) putTransactionLog(blk *block.Block) error {
+	compressData := fd.compressBlock
+	sysLog := blk.TransactionLog()
+	if sysLog == nil {
+		sysLog = &block.BlkTransactionLog{}
+		compressData = false
+	}
+	logBytes, err := compressDatabytes(sysLog.Serialize(), compressData)
+	if err != nil {
+		return err
+	}
+	return addOneEntryToBatch(fd.sysStore, logBytes, fd.batch)
+}
+
+func addOneEntryToBatch(c db.CountingIndex, v []byte, b batch.KVStoreBatch) error {
+	if err := c.UseBatch(b); err != nil {
+		return err
+	}
+	if err := c.Add(v, true); err != nil {
+		return err
+	}
+	return c.Finalize()
+}
+
+func compressDatabytes(value []byte, comp bool) ([]byte, error) {
+	if value == nil {
+		return nil, ErrDataCorruption
+	}
+
+	// append 1 byte to end of bytestream to indicate the block is compressed or not
+	if !comp {
+		return append(value, _normal), nil
+	}
+
+	var err error
+	value, err = compress.CompV2(value)
+	if err != nil {
+		return nil, err
+	}
+	return append(value, _compressed), nil
+}
+
+func decompressDatabytes(value []byte) ([]byte, error) {
+	if len(value) == 0 {
+		return nil, ErrDataCorruption
+	}
+
+	// last byte of bytestream indicates the block is compressed or not
+	switch value[len(value)-1] {
+	case _normal:
+		return value[:len(value)-1], nil
+	case _compressed:
+		return compress.DecompV2(value[:len(value)-1])
+	default:
+		return nil, ErrDataCorruption
+	}
+}
+
+func getValueMustBe8Bytes(kvs db.KVStore, ns string, key []byte) ([]byte, error) {
+	value, err := kvs.Get(ns, key)
+	if err != nil {
+		return nil, err
+	}
+	if len(value) != 8 {
+		return nil, ErrDataCorruption
+	}
+	return value, nil
+}
+
+// blockStoreKey is the slot of block in block storage (each item containing blockStorageBatchSize of blocks)
+func blockStoreKey(height, start uint64) uint64 {
+	if height <= start {
+		return 0
+	}
+	return (height - start) / blockStoreBatchSize
+}
+
+// stagingKey is the position of block in the staging buffer
+func stagingKey(height, start uint64) uint64 {
+	return (height - start) % blockStoreBatchSize
+}
```

Comment:
i am not sure if this is better or let stagingbuff maintain next index is better. I actually prefer later one, but i cannot tell exactly why.